src/storage/src/hummock/cache.rs

```diff
@@ -0,0 +1,692 @@
+// Copyright 2022 Singularity Data
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+// http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+use std::ptr::null_mut;
+use std::sync::Arc;
+
+use spin::Mutex;
+
+const IN_CACHE: u8 = 1;
+const REVERSE_IN_CACHE: u8 = !IN_CACHE;
+// LRU cache implementation port from github.com/facebook/rocksdb. The class `LruCache` is
+// thread-safe, because every operation on cache will be protected by a spin lock.
+
+// An entry is a variable length heap-allocated structure.
+// Entries are referenced by cache and/or by any external entity.
+// The cache keeps all its entries in a hash table. Some elements
+// are also stored on LRU list.
+//
+// LruHandle can be in these states:
+// 1. Referenced externally AND in hash table.
+//    In that case the entry is *not* in the LRU list
+//    (refs >= 1 && in_cache == true)
+// 2. Not referenced externally AND in hash table.
+//    In that case the entry is in the LRU list and can be freed.
+//    (refs == 0 && in_cache == true)
+// 3. Referenced externally AND not in hash table.
+//    In that case the entry is not in the LRU list and not in hash table.
+//    The entry can be freed when refs becomes 0.
+//    (refs >= 1 && in_cache == false)
+//
+// All newly created LruHandles are in state 1. If you call
+// LruCacheShard::Release on entry in state 1, it will go into state 2.
+// To move from state 1 to state 3, either call LruCacheShard::Erase or
+// LruCacheShard::Insert with the same key (but possibly different value).
+// To move from state 2 to state 1, use LruCacheShard::Lookup.
+// Before destruction, make sure that no handles are in state 1. This means
+// that any successful LruCacheShard::Lookup/LruCacheShard::Insert have a
+// matching LruCache::Release (to move into state 2) or LruCacheShard::Erase
+// (to move into state 3).
+pub struct LruHandle<K: PartialEq + Default, T> {
+    next_hash: *mut LruHandle<K, T>,
```

Comment:
better to doc the fields.

next_hash: next element in the hash bucket the LRU handle represents
next: next element in LRU linked list
prev: prev element in LRU linked list
etc.